### DISCUSSION TITLE
CompiledBinding  bind nullable property to not nullable

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#13555] [Q&A]
While developing a control, I encountered strange behavior when binding the nullable property to non-nullable. 
Considering the following fragment:

```c#
    public class MyControl: Control
    {
        public readonly static StyledProperty<decimal?> SelectedValueProperty =
            AvaloniaProperty.Register<MyControl, decimal?>(nameof(SelectedValue),defaultBindingMode:BindingMode.TwoWay);

        public decimal? SelectedValue 
        { 
            get => GetValue(SelectedValueProperty);
            set => SetValue(SelectedValueProperty, value);
        }
    }
``` 

```c#
                    var xaml = $@"
<Window xmlns='https://github.com/avaloniaui'
        xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'
        xmlns:local='clr-namespace:Avalonia.Markup.Xaml.UnitTests.MarkupExtensions;assembly=Avalonia.Markup.Xaml.UnitTests'
        x:DataType='local:TestDataContext'
        x:CompileBindings='{true}'>
    <local:MyControl Name='my' SelectedValue='{{Binding DecimalValue}}'/>
</Window>";
                    var window = (Window)AvaloniaRuntimeXamlLoader.Load(xaml);
                    var my = window.FindControl<MyControl>("my");

                    var dataContext = new TestDataContext();
                    window.DataContext = dataContext;

                    Assert.Equal(TestDataContext.ExpectedDecimal, my.GetValue(MyControl.SelectedValueProperty));


                    my.ClearValue(MyControl.SelectedValueProperty);

                    Assert.Null(my.SelectedValue);

                    Assert.Equal(0m, dataContext.DecimalValue);
``` 

if I invoke `my.ClearValue(MyControl.SelectedValueProperty);` I expected `dataContext.DecimalValue` to be `0` but it actually hasn't changed and there is nothing in the log and any exception is throwed.

I resolved using the Converter, but not i sure if actual behavior is correct. Searching i found simile issue with NumericUpDown #10793.

I stepped into this behavior and I've seen that when

- invoke `my.ClearValue(MyControl.SelectedValueProperty);`  will be restored to default value (`null`).
- Binding invoke`ClrPropertyInfo.Set` thath throw `NullReferenceException` because `TestDataContext.DecimalValue` is not-nullable 

 https://github.com/AvaloniaUI/Avalonia/blob/9cd8dda5206ffbcb6705ada2ae8d3753aaaf9caf/src/Avalonia.Base/Data/Core/ClrPropertyInfo.cs#L34

- NullReferenceException  is trapped form `PropertyAccessorNode.SetTargetValueCore` 

https://github.com/AvaloniaUI/Avalonia/blob/9cd8dda5206ffbcb6705ada2ae8d3753aaaf9caf/src/Avalonia.Base/Data/Core/PropertyAccessorNode.cs#L31-L44

- No exceptions and the log is generated

I don't know what the right behavior is, whether to set the non-nullable property to the default value or keep the current one.
The only thing that in my opinion is necessary is to inform the developer of the problem in some way (exception, log entry, etc.).

I won't propose any PR because I remember that the team are working on refactoring the Binding Engine.

cc @grokys

**Comment by timunie** [comment]
yes, we are. there is a branch for it but no PR to test yet. It's not production ready yet.
